module/zfs/zpl_xattr.c

```diff
@@ -478,8 +478,11 @@ zpl_xattr_set_sa(struct inode *ip, const char *name, const void *value,
 	}
 
 	/* Update the SA for additions, modifications, and removals. */
-	if (!error)
+	if (!error) {
 		error = -zfs_sa_set_xattr(zp);
+		if (error && (value != NULL))
+			(void) nvlist_remove(nvl, name, DATA_TYPE_BYTE_ARRAY);
```

Comment:
My thinking was that we should only panic the system if proceeding would result in pool corruption or returning bad data. I'm not sure that applies here. But perhaps the better thing to do would be to invalidate `zp->z_xattr_cached` if `nvlist_remove()` fails.